lib/Parse/ParseExpr.cpp

```diff
@@ -2905,9 +2910,67 @@ ExprResult Parser::ParseBoundsExpression() {
     SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
 
   PT.consumeClose();
+
+  if (StartsRelativeBoundsClause(Tok) && !Result.isInvalid()) {
+    if (ParseRelativeBoundsClause(Result.get())) {
+      Diag(Tok, diag::err_unexpected_return_relative_bounds_clause);
+      Result = ExprError();
+    }
+  }
+
   return Result;
 }
 
+bool Parser::ParseRelativeBoundsClause(Expr * Expr) {
+  bool IsError = false;
+  RelativeBoundsClause *RelativeClause = nullptr;
+  ExprResult ConstExpr;
+
+  IdentifierInfo *Ident = Tok.getIdentifierInfo();
+  SourceLocation BoundsKWLoc = Tok.getLocation();
+  ConsumeToken();
+
+  // Skip until right paren and do nothing if it is not range bounds expr.
+  // for example bounds(none, none + 5). Currently, it is NullaryBoundsExpr
+  if (!(cast<BoundsExpr>(Expr)->isRange())) {
```

Comment:
You'll be able to delete this error recover  code if you make the change to ParseBoundsExpression that I recommend.